pallets/delegating/src/lib.rs

```diff
@@ -0,0 +1,494 @@
+
+        pub ErasValidatorReward get(fn eras_validator_reward):
+            map hasher(twox_64_concat) EraIndex => Option<BalanceOf<T>>;
+
+        // (EraIndex, validatorId) -> Vec<(delegator, score, hasSlashed)>
+        SelectedDelegators get(fn selected_delegators): double_map hasher(blake2_128_concat) EraIndex,
+        hasher(blake2_128_concat) T::AccountId => Vec<(T::AccountId, u64, bool)>;
+
+        pub CurrentEra get(fn current_era): Option<EraIndex>;
+        pub CurrentEraValidators get(fn current_era_validators): Option<Vec<T::AccountId>>;
+
+        //  candidate validator list
+        pub CandidateValidators get(fn get_candidate_validators): Option<Vec<T::AccountId>>;
+
+    }
+}
+
+// Pallets use events to inform users when important changes are made.
+// https://substrate.dev/docs/en/knowledgebase/runtime/events
+decl_event!(
+    pub enum Event<T>
+    where
+        AccountId = <T as frame_system::Trait>::AccountId,
+        Balance = BalanceOf<T>,
+    {
+        /// Event documentation should end with an array that provides descriptive names for event
+        /// parameters. [something, who]
+        Delegated(AccountId),
+        UnDelegated(AccountId),
+        WithdrawCredit(AccountId, u64),
+        /// The staker has been rewarded by this amount. \[stash, amount\]
+        Reward(AccountId, Balance),
+    }
+);
+
+// Errors inform users that something went wrong.
+decl_error! {
+    pub enum Error for Module<T: Trait> {
+        /// Error names should be descriptive.
+        NoneValue,
+        /// Errors should have helpful documentation associated with them.
+        StorageOverflow,
+        AlreadyDelegated,
+        NotDelegate,
+        CreditLocked,
+        NoCreditLedgerData,
+        NotRightEra,
+        CreditScoreTooLow,
+        NonCandidateValidator,
+        NotInCandidateValidator,
+        SelectTooManyValidators,
+        SelectNoValidator,
+        DelegateMoreCreditScoreThanOwn,
+        InvalidEraToReward,
+    }
+}
+
+// Dispatchable functions allows users to interact with the pallet and invoke state changes.
+// These functions materialize as "extrinsics", which are often compared to transactions.
+// Dispatchable functions must be annotated with a weight and must return a DispatchResult.
+decl_module! {
+    pub struct Module<T: Trait> for enum Call where origin: T::Origin {
+        // Errors must be initialized if they are used by the pallet.
+        type Error = Error<T>;
+
+        // Events must be initialized if they are used by the pallet.
+        fn deposit_event() = default;
+
+        /// delegate credit score to validators in vec equally
+        #[weight = 10_000]
+        pub fn delegate(origin, validators: Vec<T::AccountId>) -> dispatch::DispatchResult {
+            // check signature
+            let controller = ensure_signed(origin)?;
+
+            // check credit pass threshold
+            if T::CreditInterface::pass_threshold(controller.clone(), 0) == false{
```

Comment:
what does the second parameter _ttype mean, and how is it going to used in future?

Reply:
In page [https://www.notion.so/deeper-chain-testnet-plan-ac96c69b36374e419dd0324662f517f5](https://www.notion.so/deeper-chain-testnet-plan-ac96c69b36374e419dd0324662f517f5),  `pub fn pass_threshold(origin, ttype: u8) -> DispatchResult{}`

Reply:
got it. it's placeholder for future threshold type